primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs

```diff
@@ -91,20 +92,46 @@ fn function_for_method(
 }
 
 /// Generates the bare function implementation for `cfg(not(feature = "std"))`.
-fn function_no_std_impl(method: &TraitItemMethod) -> Result<TokenStream> {
+fn function_no_std_impl(
+	method: &RuntimeInterfaceFunction,
+	is_wasm_only: bool,
+) -> Result<TokenStream> {
 	let function_name = &method.sig.ident;
 	let host_function_name = create_exchangeable_host_function_ident(&method.sig.ident);
 	let args = get_function_arguments(&method.sig);
 	let arg_names = get_function_argument_names(&method.sig);
-	let return_value = &method.sig.output;
+	let return_value = if method.should_trap_on_return() {
+		syn::ReturnType::Type(
+			<Token![->]>::default(),
+			Box::new(syn::TypeNever { bang_token: <Token![!]>::default() }.into()),
+		)
```

Comment:
Why not require that the declaration directly uses `!` as return type? Would make it less "magic".

If not, we could also use here `parse_quote!( -> ! )`?

Reply:
I actually considered that too, but then it kinda becomes magic in a different way... (:

For example:

```rust
fn foobar(&mut self) -> ! {
    do_stuff();
}
```

...if this was a normal Rust function it wouldn't compile because it returns `()` and *not* `!`. So in a way it's confusing because the function is then *not* supposed to actually return what its prototype says it returns! (Since the body is executed *on the host's side* while the `core::arch::wasm32::unreachable` is executed *on the WASM's side*.)

Alternatively to resolve this inconsistency we could:

a) Require that the host function actually *does* return `!`; that'd probably require a `panic!` inside, and for us to catch it at the boundary, but that seems kinda.... silly? (And also, `!` doesn't necessarily mean panic; it can also mean e.g. `std::process::exit` so conceptually it doesn't perfectly translate to the concept of a trap.)
b) Require that the host function returns a zero-sized sentinel type instead of `!`, but that just seems like more work for not much gain.

So in the end I went the attribute route. Not sure if there's another better alternative that I'm missing?

Reply:
> ...if this was a normal Rust function it wouldn't compile because it returns `()` and _not_ `!`

Good argument. Whatever we do, it will be a little bit weird :D Fine to keep it the current way.